File: Recursion and Backtracking/Damerau_Levenshtein_Distance_Recursion/Brute_Force.java
import java.util.*;

class DLResult {
    int cost;
    List<String> operations;
    DLResult() { this.cost = 0; this.operations = new ArrayList<>(); }
    DLResult(int cost, List<String> operations) { this.cost = cost; this.operations = operations; }
}

class OperationCosts {
    int insertCost, deleteCost, replaceCost, transposeCost;  
    OperationCosts(int ins, int del, int rep, int trans) { 
        this.insertCost = ins; this.deleteCost = del; this.replaceCost = rep; this.transposeCost = trans;  
    }
    OperationCosts() { this(1,1,1,1); }
}

public class Main {
    private static int minimumCost(int a, int b, int c) { return Math.min(a, Math.min(b, c)); }
    private static int minimumCost(int a, int b, int c, int d) { return Math.min(Math.min(a, b), Math.min(c, d)); }

    private static int damerauLevenshteinCostOnly(String str1, String str2, int i, int j, OperationCosts costs) {
        if (i == str1.length()) return (str2.length() - j) * costs.insertCost;
        if (j == str2.length()) return (str1.length() - i) * costs.deleteCost;
        if (str1.charAt(i) == str2.charAt(j)) return damerauLevenshteinCostOnly(str1, str2, i + 1, j + 1, costs);
        int insertCost = costs.insertCost + damerauLevenshteinCostOnly(str1, str2, i, j + 1, costs);
        int deleteCost = costs.deleteCost + damerauLevenshteinCostOnly(str1, str2, i + 1, j, costs);
        int replaceCost = costs.replaceCost + damerauLevenshteinCostOnly(str1, str2, i + 1, j + 1, costs);
        if (i + 1 < str1.length() && j + 1 < str2.length() && str1.charAt(i) == str2.charAt(j + 1) && str1.charAt(i + 1) == str2.charAt(j)) {
            int transposeCost = costs.transposeCost + damerauLevenshteinCostOnly(str1, str2, i + 2, j + 2, costs);
            return minimumCost(insertCost, deleteCost, replaceCost, transposeCost);
        }
        return minimumCost(insertCost, deleteCost, replaceCost);
    }

    private static DLResult damerauLevenshteinCostAndOperation(String str1, String str2, int i, int j, OperationCosts costs) {
        if (i == str1.length()) {
            List<String> ops = new ArrayList<>(); int totalCost = 0;
            for (int k = j; k < str2.length(); k++) { ops.add("Insert '" + str2.charAt(k) + "'"); totalCost += costs.insertCost; }
            return new DLResult(totalCost, ops);
        }
        if (j == str2.length()) {
            List<String> ops = new ArrayList<>(); int totalCost = 0;
            for (int k = i; k < str1.length(); k++) { ops.add("Delete '" + str1.charAt(k) + "'"); totalCost += costs.deleteCost; }
            return new DLResult(totalCost, ops);
        }
        if (str1.charAt(i) == str2.charAt(j)) {
            DLResult res = damerauLevenshteinCostAndOperation(str1, str2, i + 1, j + 1, costs);
            res.operations.add(0, "Match '" + str1.charAt(i) + "'"); return res;
        }
        DLResult insertOp = damerauLevenshteinCostAndOperation(str1, str2, i, j + 1, costs);
        insertOp.cost += costs.insertCost; insertOp.operations.add(0, "Insert '" + str2.charAt(j) + "'");
        DLResult deleteOp = damerauLevenshteinCostAndOperation(str1, str2, i + 1, j, costs);
        deleteOp.cost += costs.deleteCost; deleteOp.operations.add(0, "Delete '" + str1.charAt(i) + "'");
        DLResult replaceOp = damerauLevenshteinCostAndOperation(str1, str2, i + 1, j + 1, costs);
        replaceOp.cost += costs.replaceCost; replaceOp.operations.add(0, "Replace '" + str1.charAt(i) + "' with '" + str2.charAt(j) + "'");
        if (i + 1 < str1.length() && j + 1 < str2.length() && str1.charAt(i) == str2.charAt(j + 1) && str1.charAt(i + 1) == str2.charAt(j)) {
            DLResult transposeOp = damerauLevenshteinCostAndOperation(str1, str2, i + 2, j + 2, costs);
            transposeOp.cost += costs.transposeCost; transposeOp.operations.add(0, "Transpose '" + str1.charAt(i) + str1.charAt(i + 1) + "'");
            if (insertOp.cost <= deleteOp.cost && insertOp.cost <= replaceOp.cost && insertOp.cost <= transposeOp.cost) return insertOp;
            else if (deleteOp.cost <= insertOp.cost && deleteOp.cost <= replaceOp.cost && deleteOp.cost <= transposeOp.cost) return deleteOp;
            else if (replaceOp.cost <= insertOp.cost && replaceOp.cost <= deleteOp.cost && replaceOp.cost <= transposeOp.cost) return replaceOp;
            else return transposeOp;
        }
        if (insertOp.cost <= deleteOp.cost && insertOp.cost <= replaceOp.cost) return insertOp;
        else if (deleteOp.cost <= insertOp.cost && deleteOp.cost <= replaceOp.cost) return deleteOp;
        else return replaceOp;
    }

    public static void main(String[] args) {
        OperationCosts costs = new OperationCosts(1,1,1,1);
        String s1 = "kitten", s2 = "sitting";
        System.out.println("Edit Distance: " + damerauLevenshteinCostOnly(s1, s2, 0, 0, costs));
        DLResult res = damerauLevenshteinCostAndOperation(s1, s2, 0, 0, costs);
        System.out.println("Operations:"); for (String op : res.operations) System.out.println(op);
    }
}

File: Recursion and Backtracking/Damerau_Levenshtein_Distance_Recursion/Brute_Force.cpp
#include<iostream>
#include<vector>
#include<string>
using namespace std;

struct DL_Result {
    int cost;
    vector<string> operations;
    DL_Result() : cost(0), operations({}) {}
    DL_Result(int c, vector<string> ops) : cost(c), operations(ops) {}
};

struct OperationCosts {
    int insert_cost, delete_cost, replace_cost, transpose_cost; 
    OperationCosts(int ins=1, int del=1, int rep=1, int trans=1) 
        : insert_cost(ins), delete_cost(del), replace_cost(rep), transpose_cost(trans) {}
};

int minimum_cost(int a, int b, int c) { return min(a, min(b, c)); }
int minimum_cost(int a, int b, int c, int d) { return min(min(a, b), min(c, d)); }

int damerauLevenshtein_Cost_Only(const string &str1, const string &str2, int i, int j, const OperationCosts &costs) {
    if (i == str1.length()) return (str2.length() - j) * costs.insert_cost;
    if (j == str2.length()) return (str1.length() - i) * costs.delete_cost;
    if (str1[i] == str2[j]) return damerauLevenshtein_Cost_Only(str1, str2, i + 1, j + 1, costs);
    int insertCost = costs.insert_cost + damerauLevenshtein_Cost_Only(str1, str2, i, j + 1, costs);
    int deleteCost = costs.delete_cost + damerauLevenshtein_Cost_Only(str1, str2, i + 1, j, costs);
    int replaceCost = costs.replace_cost + damerauLevenshtein_Cost_Only(str1, str2, i + 1, j + 1, costs);
    if (i + 1 < str1.length() && j + 1 < str2.length() && str1[i] == str2[j + 1] && str1[i + 1] == str2[j]) {
        int transposeCost = costs.transpose_cost + damerauLevenshtein_Cost_Only(str1, str2, i + 2, j + 2, costs);
        return minimum_cost(insertCost, deleteCost, replaceCost, transposeCost);
    }
    return minimum_cost(insertCost, deleteCost, replaceCost);
}

DL_Result damerauLevenshtein_Cost_And_Operation(string str1, string str2, int i, int j, const OperationCosts &costs) {
    if (i == str1.length()) {
        vector<string> ops; int total_cost = 0;
        for (int k = j; k < str2.length(); k++) {
            ops.push_back("Insert '" + string(1, str2[k]) + "'");
            total_cost += costs.insert_cost;
        }
        return DL_Result(total_cost, ops);
    }
    if (j == str2.length()) {
        vector<string> ops; int total_cost = 0;
        for (int k = i; k < str1.length(); k++) {
            ops.push_back("Delete '" + string(1, str1[k]) + "'");
            total_cost += costs.delete_cost;
        }
        return DL_Result(total_cost, ops);
    }
    if (str1[i] == str2[j]) {
        DL_Result res = damerauLevenshtein_Cost_And_Operation(str1, str2, i + 1, j + 1, costs);
        res.operations.insert(res.operations.begin(), "Match '" + string(1, str1[i]) + "'");
        return res;
    }
    DL_Result insert_Op = damerauLevenshtein_Cost_And_Operation(str1, str2, i, j + 1, costs);
    insert_Op.cost += costs.insert_cost;
    insert_Op.operations.insert(insert_Op.operations.begin(), "Insert '" + string(1, str2[j]) + "'");
    DL_Result delete_Op = damerauLevenshtein_Cost_And_Operation(str1, str2, i + 1, j, costs);
    delete_Op.cost += costs.delete_cost;
    delete_Op.operations.insert(delete_Op.operations.begin(), "Delete '" + string(1, str1[i]) + "'");
    DL_Result replace_Op = damerauLevenshtein_Cost_And_Operation(str1, str2, i + 1, j + 1, costs);
    replace_Op.cost += costs.replace_cost;
    replace_Op.operations.insert(replace_Op.operations.begin(), "Replace '" + string(1, str1[i]) + "' with '" + string(1, str2[j]) + "'");
    if (i + 1 < str1.length() && j + 1 < str2.length() && str1[i] == str2[j + 1] && str1[i + 1] == str2[j]) {
        DL_Result transpose_Op = damerauLevenshtein_Cost_And_Operation(str1, str2, i + 2, j + 2, costs);
        transpose_Op.cost += costs.transpose_cost;
        transpose_Op.operations.insert(transpose_Op.operations.begin(), "Transpose '" + string(1, str1[i]) + string(1, str1[i + 1]) + "'");
        if (insert_Op.cost <= delete_Op.cost && insert_Op.cost <= replace_Op.cost && insert_Op.cost <= transpose_Op.cost) return insert_Op;
        else if (delete_Op.cost <= insert_Op.cost && delete_Op.cost <= replace_Op.cost && delete_Op.cost <= transpose_Op.cost) return delete_Op;
        else if (replace_Op.cost <= insert_Op.cost && replace_Op.cost <= delete_Op.cost && replace_Op.cost <= transpose_Op.cost) return replace_Op;
        else return transpose_Op;
    }
    if (insert_Op.cost <= delete_Op.cost && insert_Op.cost <= replace_Op.cost) return insert_Op;
    else if (delete_Op.cost <= insert_Op.cost && delete_Op.cost <= replace_Op.cost) return delete_Op;
    else return replace_Op;
}

int main() {
    OperationCosts costs(1,2,3,1);
    string s1 = "kitten", s2 = "sitting";
    cout << "Edit Distance: " << damerauLevenshtein_Cost_Only(s1, s2, 0, 0, costs) << "\n";
    DL_Result res = damerauLevenshtein_Cost_And_Operation(s1, s2, 0, 0, costs);
    cout << "Operations:\n"; 
    for (auto &op : res.operations) cout << op << "\n";
}
